pkg/connector/ble/ble.go

```diff
@@ -248,54 +328,83 @@ func tryToConnect(ctx context.Context, vin string, target Advertisement) (*Conne
+	log.Debug("Connecting to %s (%s)...", target.Address.String(), localName)
+
+	// FIXME: This is a workaround for the fact that bluetooth library doesn't
+	// support context handling. While not a big issue, it will be good to
+	// have this fixed upstream. Also applies to the DiscoverServices and
+	// DiscoverCharacteristics calls below. See:
+	// https://github.com/tinygo-org/bluetooth/issues/339
+	deviceCh := make(chan bluetooth.Device)
+	errorCh := make(chan error)
+	connectionCancelled := false
+	go func() {
+		params := bluetooth.ConnectionParams{}
+		// If a deadline is set, use it as the connection timeout
+		// else this go routine will expire after the default timeout.
+		if deadline, ok := ctx.Deadline(); ok {
+			params.ConnectionTimeout = bluetooth.NewDuration(time.Until(deadline))
+		}
+		device, err := adapter.Connect(target.Address, params)
+		if err != nil && !connectionCancelled {
+			errorCh <- err
+		} else if !connectionCancelled {
+			deviceCh <- device
+		} else {
+			if err := device.Disconnect(); err != nil {
+				log.Warning("ble: failed to disconnect: %s", err)
+			}
+		}
+	}()
+	var device bluetooth.Device
+	select {
+	case device = <-deviceCh:
+		log.Debug("Connected to %s", target.Address.String())
+	case err := <-errorCh:
+		return nil, true, fmt.Errorf("ble: failed to connect to device: %s", err)
+	case <-ctx.Done():
+		connectionCancelled = true
+		return nil, true, ctx.Err()
 	}
 
-	log.Debug("Discovering services %s...", client.Addr())
-	services, err := client.DiscoverServices([]ble.UUID{vehicleServiceUUID})
+	log.Debug("Discovering services %s...", device.Address.String())
+	services, err := device.DiscoverServices([]bluetooth.UUID{vehicleServiceUUID})
 	if err != nil {
 		return nil, true, fmt.Errorf("ble: failed to enumerate device services: %s", err)
 	}
-	if len(services) == 0 {
+	if len(services) != 1 {
 		return nil, true, fmt.Errorf("ble: failed to discover service")
 	}
 
-	characteristics, err := client.DiscoverCharacteristics([]ble.UUID{toVehicleUUID, fromVehicleUUID}, services[0])
+	characteristics, err := services[0].DiscoverCharacteristics([]bluetooth.UUID{toVehicleUUID, fromVehicleUUID})
 	if err != nil {
 		return nil, true, fmt.Errorf("ble: failed to discover service characteristics: %s", err)
 	}
 
-	conn := Connection{
-		vin:    vin,
-		client: client,
-		inbox:  make(chan []byte, 5),
+	if len(characteristics) != 2 {
+		return nil, true, errors.New("ble: failed to find required characteristics")
 	}
-	for _, characteristic := range characteristics {
-		if characteristic.UUID.Equal(toVehicleUUID) {
-			conn.txChar = characteristic
-		} else if characteristic.UUID.Equal(fromVehicleUUID) {
-			conn.rxChar = characteristic
-		}
-		if _, err := client.DiscoverDescriptors(nil, characteristic); err != nil {
-			return nil, true, fmt.Errorf("ble: couldn't fetch descriptors: %s", err)
-		}
+
+	mtu, err := characteristics[0].GetMTU()
```

Comment:
The MTU of the BLE endpoint may exceed the available memory allocated by the vehicle's security controller. We'll need to take the smaller of the two. I'll follow up what the smallest value this is across the fleet.

Reply:
Can't we parse vehicle model of of the VIN? I know its possible to determine model/year.

Reply:
Disregard this comment, confirmed with a developer that using the BLEEP MTU should be fine.